### PR TITLE
platforms/joystick: replace sdl with 0xcafed00d/joystick  package

### PR DESCRIPTION
**Pull request by deadprogram**
## Solved issues and/or description of the change

This PR is to address https://github.com/hybridgroup/gobot/issues/986 by replacing SDL with the much simpler package https://github.com/0xcafed00d/joystick

## Manual test

Already tested Dualshock 3 on both Linux and macOS. Will test Windows this weekend.

## Checklist

- [X] The PR's target branch is 'hybridgroup:dev'
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes (e.g. by run `make test`)
- [X] No linter errors exist locally (e.g. by run `make fmt_check`)
- [X] I have performed a self-review of my own code

### CODE REVIEW

File: platforms/joystick/joystick_adaptor.go [hunk 1]
```diff
@@ -1,38 +1,35 @@
...
-	joystick joystick
+	id       int
+	joystick js.Joystick
 	connect  func(*Adaptor) error
 }
 
 // NewAdaptor returns a new Joystick Adaptor.
-func NewAdaptor() *Adaptor {
+// Pass in the ID of the joystick you wish to connect to.
+func NewAdaptor(id int) *Adaptor {
```
Comment (gen2thomas) [inline platforms/joystick/joystick_adaptor.go]: if we make "id" a string, we are able to implement differentiation by name later [see comment in 564](https://github.com/hybridgroup/gobot/issues/564#issuecomment-1254040934) and its different implementation ideas.

File: platforms/joystick/joystick_adaptor_test.go [hunk 2]
```diff
@@ -31,8 +30,9 @@ func TestAdaptorConnect(t *testing.T) {
 	a := initTestAdaptor()
 	gobottest.Assert(t, a.Connect(), nil)
```
Comment (gen2thomas) [inline platforms/joystick/joystick_adaptor_test.go]:
```suggestion
	assert.Nil(t, a.Connect())
```
my suggestion is to drop usage of gobottest in this small file and use "github.com/stretchr/testify/assert" instead, see #960

File: platforms/joystick/joystick_adaptor_test.go [hunk 3]
```diff
@@ -31,8 +30,9 @@ func TestAdaptorConnect(t *testing.T) {
...
 
-	a = NewAdaptor()
-	gobottest.Assert(t, a.Connect(), errors.New("No joystick available"))
+	a = NewAdaptor(6)
+	err := a.Connect()
+	gobottest.Assert(t, strings.HasPrefix(err.Error(), "No joystick available"), true)
```
Comment (gen2thomas) [inline platforms/joystick/joystick_adaptor_test.go]:
```suggestion
	assert.True(t, strings.HasPrefix(err.Error(), "No joystick available"))
```

File: platforms/joystick/joystick_driver.go [hunk 4]
```diff
@@ -126,6 +120,41 @@ func (j *Driver) adaptor() *Adaptor {
 //		[button]_release
 //		[axis]
 func (j *Driver) Start() (err error) {
```
Comment (gen2thomas) [inline platforms/joystick/joystick_driver.go]:
```suggestion
func (j *Driver) Start() error {
```

File: platforms/joystick/joystick_driver.go [hunk 5]
```diff
@@ -187,48 +200,48 @@ func (j *Driver) Halt() (err error) {
 	return
 }
 
-var previousHat = ""
+func (j *Driver) handleButtons(state js.State) error {
```
Comment (gen2thomas) [inline platforms/joystick/joystick_driver.go:204]: I'm not able to make a suggestion directly for this function, so I try this way:
```
func (j *Driver) handleButtons(state js.State) error {
	for button := 0; button < j.adaptor().joystick.ButtonCount(); button++ {
		buttonPressed := state.Buttons&(1<<uint32(button)) != 0
		if buttonPressed != j.buttonState[button] {
			j.buttonState[button] = buttonPressed
			name := j.findName(uint8(button), j.config.Buttons)
			if name == "" {
				return fmt.Errorf("Unknown button: %v", button)
			}

			if buttonPressed {
				j.Publish(j.Event(fmt.Sprintf("%s_press", name)), nil)
			} else {
				j.Publish(j.Event(fmt.Sprintf("%s_release", name)), nil)
			}
		}
	}

	return nil
}
```

File: examples/joystick_ps3.go [hunk 6]
```diff
@@ -92,15 +104,27 @@ func main() {
 		stick.On(joystick.R1Press, func(data interface{}) {
 			fmt.Println("R1Press", data)
```
Comment (gen2thomas) [inline examples/joystick_ps3.go:105]:
```suggestion
			fmt.Println("R1_press", data)
```

File: examples/joystick_ps3.go [hunk 7]
```diff
@@ -92,15 +104,27 @@ func main() {
...
 		})
+		stick.On(joystick.R1Release, func(data interface{}) {
+			fmt.Println("R1Release", data)
```
Comment (gen2thomas) [inline examples/joystick_ps3.go:108]:
```suggestion
			fmt.Println("R1_release", data)
```

File: examples/joystick_ps3.go [hunk 8]
```diff
@@ -92,15 +104,27 @@ func main() {
...
+		})
 		stick.On(joystick.R2Press, func(data interface{}) {
 			fmt.Println("R2Press", data)
```
Comment (gen2thomas) [inline examples/joystick_ps3.go:111]:
```suggestion
			fmt.Println("R2_press", data)
```

File: examples/joystick_ps3.go [hunk 9]
```diff
@@ -92,15 +104,27 @@ func main() {
...
 		})
+		stick.On(joystick.R2Release, func(data interface{}) {
+			fmt.Println("R2Release", data)
```
Comment (gen2thomas) [inline examples/joystick_ps3.go:114]:
```suggestion
			fmt.Println("R2_release", data)
```

File: examples/joystick_ps3.go [hunk 10]
```diff
@@ -92,15 +104,27 @@ func main() {
...
+		})
 		stick.On(joystick.L1Press, func(data interface{}) {
 			fmt.Println("L1Press", data)
```
Comment (gen2thomas) [inline examples/joystick_ps3.go:117]:
```suggestion
			fmt.Println("L1_press", data)
```

File: examples/joystick_ps3.go [hunk 11]
```diff
@@ -92,15 +104,27 @@ func main() {
...
 		})
+		stick.On(joystick.L1Release, func(data interface{}) {
+			fmt.Println("L1Release", data)
```
Comment (gen2thomas) [inline examples/joystick_ps3.go:120]:
```suggestion
			fmt.Println("L1_release", data)
```

File: examples/joystick_ps3.go [hunk 12]
```diff
@@ -92,15 +104,27 @@ func main() {
...
+		})
 		stick.On(joystick.L2Press, func(data interface{}) {
 			fmt.Println("L2Press", data)
```
Comment (gen2thomas) [inline examples/joystick_ps3.go:123]:
```suggestion
			fmt.Println("L2_press", data)
```

File: examples/joystick_ps3.go [hunk 13]
```diff
@@ -92,15 +104,27 @@ func main() {
...
 		})
+		stick.On(joystick.L2Release, func(data interface{}) {
+			fmt.Println("L2Release", data)
```
Comment (gen2thomas) [inline examples/joystick_ps3.go:126]:
```suggestion
			fmt.Println("L2_release", data)
```

File: platforms/joystick/joystick_driver_test.go [hunk 14]
```diff
@@ -5,43 +5,42 @@ import (
...
 	d.SetName("NewName")
 	gobottest.Assert(t, d.Name(), "NewName")
 }
 
 func TestDriverStart(t *testing.T) {
-	d := initTestDriver("./configs/xbox360_power_a_mini_proex.json")
+	d, _ := initTestDriver("./configs/dualshock3.json")
 	d.interval = 1 * time.Millisecond
 	gobottest.Assert(t, d.Start(), nil)
 	time.Sleep(2 * time.Millisecond)
```
Comment (gen2thomas) [inline platforms/joystick/joystick_driver_test.go:40]:
```suggestion
```
IMO for parallel tests this has no effect

File: go.mod [hunk 15]
```diff
@@ -3,6 +3,7 @@ module gobot.io/x/gobot/v2
 go 1.18
 
 require (
+	github.com/0xcafed00d/joystick v1.0.1
```
Comment (gen2thomas) [inline go.mod:6]: "github.com/veandco/go-sdl2 v0.4.35" is still needed?

File: .circleci/config.yml [hunk 16]
```diff
@@ -38,10 +38,10 @@ jobs:
           name: Debug version
           command: go version
       - run:
-          # digispark needs libusb, joystick needs sdl2, opencv needs opencv
-          name: Platform tests (except digispark, joystick, opencv)
+          # digispark needs libusb, opencv needs opencv
```
Comment (gen2thomas) [inline .circleci/config.yml:41]: The "Makefile" still contains this filter for examples. Maybe we can remove it also there.

File: platforms/joystick/bin/scanner.go [hunk 17]
```diff
@@ -8,65 +8,106 @@
 // Based on original code from Jacky Boen
 // https://github.com/veandco/go-sdl2/blob/master/examples/events/events.go
```
Comment (gen2thomas) [inline platforms/joystick/bin/scanner.go]:
```suggestion
```
I'm not sure for this suggestion.

File: platforms/joystick/joystick_driver_test.go
Comment (gen2thomas) [inline platforms/joystick/joystick_driver_test.go:1]: Should we add a test for using the const "Dualshock3"?

File: platforms/joystick/joystick_driver_test.go [hunk 18]
```diff
@@ -50,97 +49,62 @@ func TestDriverHalt(t *testing.T) {
 
 func TestDriverHandleEvent(t *testing.T) {
 	sem := make(chan bool)
-	d := initTestDriver("./configs/xbox360_power_a_mini_proex.json")
-	_ = d.Start()
+	d, tj := initTestDriver("./configs/dualshock3.json")
+	tj.axisCount = 6
```
Comment (gen2thomas) [inline platforms/joystick/joystick_driver_test.go:110]:
```suggestion
	tj.axisCount = 6  // corresponding to the content of json config
```

File: platforms/joystick/joystick_driver.go [hunk 19]
```diff
@@ -57,20 +59,12 @@ type pair struct {
 	ID   int    `json:"id"`
 }
 
-// hat is a JSON representation of hat, name and id
```
Comment (gen2thomas) [inline platforms/joystick/joystick_driver.go:60]: Are the hat-switches now part of the buttons or are they no longer supported?

File: platforms/joystick/joystick_adaptor.go (hunk 1 above)
Comment (deadprogram) [reply platforms/joystick/joystick_adaptor.go]: Being able to identify a joystick by name is not so easy using the lightweight github.com/0xcafed00d/joystick package. https://github.com/0xcafed00d/joystick/blob/master/joystick_windows.go#L109 but probably could be done with a bunch more coding.

But making `id` a string at least prepares for some future enhancement so seems like a good idea.

File: platforms/joystick/joystick_driver.go (hunk 5 above)
Comment (deadprogram) [reply platforms/joystick/joystick_driver.go:204]: Looks like good simplification thanks @gen2thomas

File: platforms/joystick/bin/scanner.go (hunk 17 above)
Comment (deadprogram) [reply platforms/joystick/bin/scanner.go]: I meant to change the reference to https://github.com/0xcafed00d/joystick/blob/master/joysticktest/joysticktest.go which I will do.

File: platforms/joystick/joystick_driver.go (hunk 19 above)
Comment (deadprogram) [reply platforms/joystick/joystick_driver.go:60]: I would expect them to now either be buttons or joysticks depending on how the @0xcafed00d package handles this. I have not had any time yet to work on any other joysticks besides dualshock3.

File: platforms/joystick/joystick_adaptor.go (hunk 1 above)
Comment (deadprogram) [reply platforms/joystick/joystick_adaptor.go]: Done.

File: platforms/joystick/joystick_adaptor_test.go (hunk 2 above)
Comment (deadprogram) [reply platforms/joystick/joystick_adaptor_test.go]: Done.

File: platforms/joystick/joystick_driver.go (hunk 4 above)
Comment (deadprogram) [reply platforms/joystick/joystick_driver.go]: Done.

File: platforms/joystick/joystick_driver.go (hunk 5 above)
Comment (deadprogram) [reply platforms/joystick/joystick_driver.go:204]: Done.

File: platforms/joystick/bin/scanner.go (hunk 17 above)
Comment (deadprogram) [reply platforms/joystick/bin/scanner.go]: Done.

File: platforms/joystick/joystick_driver.go (hunk 19 above)
Comment (deadprogram) [reply platforms/joystick/joystick_driver.go:60]: Confirmed they are buttons now.

Comment (deadprogram) [reply platforms/joystick/joystick_driver.go:60]: CORRECTION: dualshock4 on linux, they are joysticks. dualshock4 on macOS apparently not supported.

File: .circleci/config.yml (hunk 16 above)
Comment (deadprogram) [reply .circleci/config.yml:41]: Done.

File: platforms/joystick/joystick_adaptor_test.go (hunk 3 above)
Comment (deadprogram) [reply platforms/joystick/joystick_adaptor_test.go]: Done.

File: platforms/joystick/joystick_driver_test.go
Comment (deadprogram) [reply platforms/joystick/joystick_driver_test.go:1]: Done.

**Comment by deadprogram** [comment]
I've now also added the mappings for dualshock4.

I don't have any working setup for any other joysticks then these 2 at the moment, but this should handle many people, and the new scanner will help handle the others for people who have those setups.

**Comment by gen2thomas** [comment]
@deadprogram nice work, I'm going to merge now!

**Pull request merged**